api/pkg/transformer/feast/bigtablestore/encoding.go

```diff
@@ -0,0 +1,425 @@
+		ftKey := featureTableKey{
+			project: project,
+			table:   featureTable,
+		}
+		featureValues[ftKey] = avroValues.(map[string]interface{})
+		featureTimestamps[ftKey] = cell.Timestamp.Time()
+	}
+	return featureValues, featureTimestamps, nil
+}
+
+// Decode decodes given BigTable rows into Feast feature values, based on
+// the assumption that the column values are stored as the concatenation of
+// schema reference hash (4 bytes) and feature values serialized in Avro format.
+func (e *Encoder) Decode(ctx context.Context, rows []bigtable.Row, req *feast.OnlineFeaturesRequest, entityKeys []*spec.Entity) (*feast.OnlineFeaturesResponse, error) {
+	avroValueByKey := make(map[string]map[featureTableKey]map[string]interface{})
+	timestampByKey := make(map[string]map[featureTableKey]time.Time)
+	for _, row := range rows {
+		avroValues, timestamps, err := e.decodeAvro(ctx, row, req.Project, entityKeys)
+		if err != nil {
+			return nil, err
+		}
+		avroValueByKey[row.Key()] = avroValues
+		timestampByKey[row.Key()] = timestamps
+	}
+
+	fieldValues := make([]*serving.GetOnlineFeaturesResponse_FieldValues, len(req.Entities))
+	for i, entity := range req.Entities {
+		bigtableKey, err := feastRowToBigTableKey(entity, entityKeys)
+		if err != nil {
+			return nil, err
+		}
+
+		fields := make(map[string]*types.Value)
+		status := make(map[string]serving.GetOnlineFeaturesResponse_FieldStatus)
+		for k, v := range entity {
+			fields[k] = v
+			status[k] = serving.GetOnlineFeaturesResponse_PRESENT
+		}
+		avroValues := avroValueByKey[bigtableKey]
+		timestamp := timestampByKey[bigtableKey]
+
+		if avroValues == nil {
+			for _, fr := range req.Features {
+				fields[fr] = &types.Value{}
+				status[fr] = serving.GetOnlineFeaturesResponse_NOT_FOUND
+			}
+			fieldValues[i] = &serving.GetOnlineFeaturesResponse_FieldValues{
+				Fields:   fields,
+				Statuses: status,
+			}
+			continue
+		}
+		for _, fr := range req.Features {
+			featureRef, err := ParseFeatureRef(fr)
+			if err != nil {
+				return nil, err
+			}
+			maxAge := e.featureMetadata[featureTableKey{
+				project: req.Project,
+				table:   featureRef.FeatureTable,
+			}].MaxAge
+			if maxAge != nil && maxAge.GetSeconds() > 0 && timestamp[featureTableKey{
+				project: req.Project,
+				table:   featureRef.FeatureTable,
+			}].Add(time.Duration(maxAge.GetSeconds())*time.Second).Before(time.Now()) {
+				fields[fr] = &types.Value{}
+				status[fr] = serving.GetOnlineFeaturesResponse_OUTSIDE_MAX_AGE
+				fieldValues[i] = &serving.GetOnlineFeaturesResponse_FieldValues{
+					Fields:   fields,
+					Statuses: status,
+				}
+				continue
+			}
+
+			avroValue := avroValueByKey[bigtableKey][featureTableKey{
+				project: req.Project,
+				table:   featureRef.FeatureTable,
+			}][featureRef.Feature]
+
+			featureSpec := e.featureSpecs[featureTableKey{
+				project: req.Project,
+				table:   featureRef.FeatureTable,
+			}]
+
+			featureType, err := getFeatureType(featureSpec, featureRef.Feature)
```

Comment:
`getFeatureType` implementation has a loop inside it, shall we build a dictionary instead, considering that we already within 2 loops.